evm/combine-signatures.js

```diff
@@ -0,0 +1,294 @@
+                isValidAddress: { contractAddress },
+                isKeccak256Hash: { commandID, payloadHash },
+            });
+
+            const chainId = chain.chainId;
+            const command = 'approveContractCall';
+            const params = defaultAbiCoder.encode(
+                ['string', 'string', 'address', 'bytes32', 'bytes32', 'uint256'],
+                [sourceChain, sourceAddress, contractAddress, payloadHash, HashZero, 0],
+            );
+
+            const data = defaultAbiCoder.encode(
+                ['uint256', 'bytes32[]', 'string[]', 'bytes[]'],
+                [chainId, [commandID], [command], [params]],
+            );
+
+            const dataHash = hashMessage(arrayify(keccak256(data)));
+
+            const { keyID } = await getEVMAddresses(config, chain.id, options);
+
+            printInfo('Original bytes message (pre-hash)', data);
+            printInfo('Message hash for validators to sign', dataHash);
+            printInfo('Vald sign command', `axelard vald-sign ${keyID} [validator-addr] ${dataHash}`);
+
+            break;
+        }
+
+        case 'constructBatch': {
+            const { batchData, execute } = options;
+
+            validateParameters({ isValidCalldata: { batchData } });
+
+            const { addresses: validatorAddresses, weights, threshold } = await getEVMAddresses(config, chain.id, options);
+
+            const validatorWeights = {};
+
+            validatorAddresses.forEach((address, index) => {
+                validatorWeights[address.toLowerCase()] = weights[index];
+            });
+
+            const signatures = readSignatures();
+
+            const sortedSignatures = signatures.sort((a, b) => {
+                const addressA = getAddressFromPublicKey(`0x${a.pub_key}`).toLowerCase();
+                const addressB = getAddressFromPublicKey(`0x${b.pub_key}`).toLowerCase();
+                return addressA.localeCompare(addressB);
+            });
+
+            const batchSignatures = [];
+            const checkedAddresses = [];
+
+            const expectedMessageHash = hashMessage(arrayify(keccak256(batchData)));
+
+            let prevKeyId = sortedSignatures[0].key_id;
+            let totalWeight = 0;
+
+            for (const signatureJSON of sortedSignatures) {
+                const keyId = signatureJSON.key_id;
+                const msgHash = `0x${signatureJSON.msg_hash}`;
+                const pubKey = `0x${signatureJSON.pub_key}`;
+                const signature = `0x${signatureJSON.signature}`;
+
+                validateParameters({
+                    isNonEmptyString: { keyId },
+                    isKeccak256Hash: { msgHash },
+                    isValidCalldata: { pubKey, signature },
+                });
+
+                if (prevKeyId !== keyId) {
+                    printError('Signatures do not contain consistent key IDs', keyId);
+                    return;
+                }
+
+                prevKeyId = keyId;
```

Comment:
This line is not needed as if prevKeyId was not equal to keyId then it would have entered if block and returned.

Reply:
This line is for the next iteration of the loop, it reassigns `prevKeyId` to the current `keyId` so it can be compared to the next `keyId` and ensure that all keyIds match

Reply:
Ah yeah that's true